contracts/basset_vault/src/tax_querier.rs

```diff
@@ -11,31 +11,25 @@ impl TaxInfo {
     #[cfg(not(test))]
     pub fn get_tax_for(&self, amount: Uint256) -> Uint256 {
         let tax_amount = if amount.is_zero() {
-            Decimal256::zero()
+            Uint256::zero()
         } else {
-            Decimal256::from_uint256(amount)
-                * (Decimal256::one() - Decimal256::one() / (Decimal256::one() + self.rate))
+            let rate_part = Decimal256::one() - Decimal256::one() / (Decimal256::one() + self.rate);
+            TaxInfo::uint_times_decimal_ceil(amount, rate_part)
         };
 
-        let tax_amount = TaxInfo::round_up(tax_amount);
         let tax_capped = std::cmp::min(tax_amount, self.cap);
         std::cmp::max(tax_capped, Uint256::one())
     }
 
-    fn round_up(decimal: Decimal256) -> Uint256 {
-        if decimal.is_zero() {
-            Uint256::zero()
+    fn uint_times_decimal_ceil(a: Uint256, b: Decimal256) -> Uint256 {
```

Comment:
I personally prefer `mul` in such contexts, instead of `times`. Something kind of `ceiled_mul_uint_decimal`